### PR TITLE
Implementation of the new features of AHA-HTTP-Interface.pdf V. 1.33

### PR DESCRIPTION
**Pull request by i-koch**
- Added all functions to handle the Fritz!Dect 500 color bulbs
- Updated README.md to include documentation for the changes
- Updated example.js with bulb output.txt
- Added setSwitchToggle
- Added setHkrBoost (deactivated because not yet tested)
- Added setHkrWindowOpen (deactivated because not yet tested)
- Added suggestions (as comments) for replacing getBatteryCharge and getWindowOpen because the values are available as properties now

### CODE REVIEW

File: example.js [hunk 1]
```diff
@@ -82,21 +82,54 @@ function thermostats() {
...
     }));
   });
 }
 
+// display bulb information
+function bulbs() {
+  return fritz.getBulbList().then(function(bulbs) {
+    console.log("Bulbs: " + bulbs + "\n");
+
+    return sequence(bulbs.map(function(bulb) {
```
Comment (andig) [inline example.js:105]: Where does this sequence function come from?

File: index.js [hunk 2]
```diff
@@ -35,12 +35,21 @@ function Fritz(username, password, uri) {
...
     if (!Fritz.prototype.DECTREPEATER)      { Object.defineProperty( Fritz.prototype, "DECTREPEATER",      {value: module.exports.FUNCTION_DECTREPEATER,      writable: false}); }
+    if (!Fritz.prototype.MICROFONE)         { Object.defineProperty( Fritz.prototype, "MICROFONE",         {value: module.exports.FUNCTION_MICROFONE,         writable: false}); }
+    if (!Fritz.prototype.TEMPLATE)          { Object.defineProperty( Fritz.prototype, "TEMPLATE",          {value: module.exports.FUNCTION_TEMPLATE,          writable: false}); }
+    if (!Fritz.prototype.HANFUNUNIT)        { Object.defineProperty( Fritz.prototype, "HANFUNUNIT",        {value: module.exports.FUNCTION_HANFUNUNIT,        writable: false}); }
+    if (!Fritz.prototype.SWITCHCTRL)        { Object.defineProperty( Fritz.prototype, "SWITCHCTRL",        {value: module.exports.FUNCTION_SWITCHCTRL,        writable: false}); }
+    if (!Fritz.prototype.LEVELCTRL)         { Object.defineProperty( Fritz.prototype, "LEVELCTRL",         {value: module.exports.FUNCTION_LEVELCTRL,         writable: false}); }
```
Comment (andig) [inline index.js]: nitpicking: s/CTRL/CONTROL/- no need for an abbreviation here unless these are Fritz-defined constants?

File: index.js [hunk 3]
```diff
@@ -309,6 +379,115 @@ function api2temp(param)
...
+// }
+
+function state2api(param)
+{
+    // convert the input to an allowed value
+    if (isNumeric(param))
+        return (param > 2 ? 2 : param);
+    else if ((typeof param) == "string")
+    {
+
```
Comment (andig) [inline index.js]: drop empty line

File: index.js [hunk 4]
```diff
@@ -309,6 +379,115 @@ function api2temp(param)
...
+// function api2time(param)
+// {
+//     // convert the input to a readable timestamp
+//     return 0;
+// }
+
+function state2api(param)
+{
+    // convert the input to an allowed value
+    if (isNumeric(param))
```
Comment (andig) [inline index.js]: add brackets

File: index.js [hunk 5]
```diff
@@ -309,6 +379,115 @@ function api2temp(param)
...
+        return (param > 2 ? 2 : param);
+    else if ((typeof param) == "string")
+    {
+        
+        switch(param.toUpperCase())
```
Comment (andig) [inline index.js]: opening bracket on previous line

File: index.js [hunk 6]
```diff
@@ -321,12 +500,21 @@ module.exports.MIN_TEMP = MIN_TEMP;
...
 module.exports.FUNCTION_THERMOSTAT          = 1 << 6;  // Comet DECT, Heizkostenregler
 module.exports.FUNCTION_ENERGYMETER         = 1 << 7;  // Energie Messgerät
 module.exports.FUNCTION_TEMPERATURESENSOR   = 1 << 8;  // Temperatursensor
 module.exports.FUNCTION_OUTLET              = 1 << 9;  // Schaltsteckdose
 module.exports.FUNCTION_DECTREPEATER        = 1 << 10; // AVM DECT Repeater
+module.exports.FUNCTION_MICROFONE           = 1 << 11; // Microphone
+module.exports.FUNCTION_TEMPLATE            = 1 << 12; // Template
+module.exports.FUNCTION_HANFUNUNIT          = 1 << 13; // HAN-FUN unit
+module.exports.FUNCTION_SWITCHCTRL          = 1 << 15; // Simple switch on/off
+module.exports.FUNCTION_LEVELCTRL           = 1 << 16; // level
```
Comment (andig) [inline index.js]: s/CTRL/CONTROL/

File: index.js [hunk 7]
```diff
@@ -644,6 +990,21 @@ module.exports.getBatteryCharge = function(sid, ain, options)
     });
 };
 
+/* The above should be replaced by this, even though the whole device list is queried
```
Comment (andig) [inline index.js]: agreed, could you implement the replacement?

File: index.js [hunk 8]
```diff
@@ -644,6 +990,21 @@ module.exports.getBatteryCharge = function(sid, ain, options)
...
+module.exports.getBatteryCharge = function(sid, ain, options)
+{
+    return module.exports.getDevice(sid, ain, options).then(function(device) {
+        return device.battery;
+    });
+};
+
+*/
+
+// Same here as for getBatteryCharge.
```
Comment (andig) [inline index.js]: agreed, could you implement the replacement?

File: index.js [hunk 9]
```diff
@@ -669,6 +1030,19 @@ module.exports.getWindowOpen = function(sid, ain, options)
     });
 };
 
+/* The above should be replaced by this, even though the whole device list is queried
```
Comment (andig) [inline index.js]: agreed, could you implement the replacement?

**Review by andig — CHANGES_REQUESTED** [review]

**Comment by andig** [comment]
Great comprehensive PR, thank you very much! Are you aware of any breaking changes this would introduce?

File: example.js (hunk 1 above)
Comment (i-koch) [reply example.js:105]: I'm not sure whether I understand what you mean. 
getBulbList returns a list of bulb ains which is then used just the same way like the ain list is used for the switches or thermostats. The only difference is that the innermost sequence is a list with only one api function call. A list with one element is still a list ;-)
I used more than one function in the innermost list while testing the bulb , so the structure was quite handy.

File: index.js (hunk 2 above)
Comment (i-koch) [reply index.js]: Done

File: index.js (hunk 3 above)
Comment (i-koch) [reply index.js]: Done

File: index.js (hunk 4 above)
Comment (i-koch) [reply index.js]: Done

File: index.js (hunk 5 above)
Comment (i-koch) [reply index.js]: Done

File: index.js (hunk 6 above)
Comment (i-koch) [reply index.js]: Done

File: index.js (hunk 7 above)
Comment (i-koch) [reply index.js]: Done

File: index.js (hunk 8 above)
Comment (i-koch) [reply index.js]: Done

File: index.js (hunk 9 above)
Comment (i-koch) [reply index.js]: Done

File: example.js (hunk 1 above)
Comment (andig) [reply example.js:105]: I think what I wanted to ask is what is `sequence`? Couldn't find the function anywhere- is that a js language construct that I have not seen before?

Comment (i-koch) [reply example.js:105]: Ah, ok. Well, it is defined in example.js line 18 to 24 :-)

Comment (andig) [reply example.js:105]: ahhh. wrong file ⭕. will take a look and merge asap

**Comment by andig** [comment]
Are you aware of any breaking changes this would introduce in terms e.g. of batter api? Is it known which boxes won't receive this update?

**Comment by i-koch** [comment]
> 
> 
> Are you aware of any breaking changes this would introduce in terms e.g. of batter api? Is it known which boxes won't receive this update?

The update to Fritz!OS 7.20 is currently available for 
- FRITZ!Box 7590, 7530
- FRITZ!Repeater 2400
and will be provided for current devices. 
I don't know whether last generation (or even older) devices will get updates.
But there are no breaking changes, the bulb functions are new and 7.20 only and maybe should be marked as such.
The battery status and the window open detection already existed in the AVM docs from 22.01.2019 (v. 1.23). 
I've had them in the last OS before 7.20 (6.something) so this should not break anything for users of that OS.

Just checked, and I already included the 7.20 requirement in the bulb section of the readme.md

**Comment by i-koch** [comment]
Still any reasons not to merge?

**Comment by andig** [comment]
Thank you very much!

**Comment by andig** [comment]
Published 0.11- thank you!

**Pull request merged**